aws_xray_sdk/core/plugins/ec2_plugin.py

```diff
@@ -17,16 +19,51 @@ def initialize():
     """
     global runtime_context
 
+    # Try the IMDSv2 endpoint for metadata
     try:
         runtime_context = {}
 
-        r = urlopen('http://169.254.169.254/latest/meta-data/instance-id', timeout=1)
-        runtime_context['instance_id'] = r.read().decode('utf-8')
+        # get session token with 60 seconds TTL to not have the token lying around for a long time
+        token = do_request(url=IMDS_URL + "api/token",
+                           headers={"X-aws-ec2-metadata-token-ttl-seconds": "60"},
+                           method="PUT")
+
+        # get instance-id metadata
+        runtime_context['instance_id'] = do_request(url=IMDS_URL + "meta-data/instance-id",
+                                                    headers={"X-aws-ec2-metadata-token": token},
+                                                    method="GET")
+
+        # get availability-zone metadata
+        runtime_context['availability_zone'] = do_request(url=IMDS_URL + "meta-data/placement/availability-zone",
+                                                          headers={"X-aws-ec2-metadata-token": token},
+                                                          method="GET")
+
+    except Exception as e:
+        # Falling back to IMDSv1 endpoint
+        log.warning("failed to get ec2 instance metadata from IMDSv2 due to {}. Falling back to IMDSv1".format(e))
```

Comment:
Consider making this a debug statement since, to my understanding, not all instances have v2 enabled

Reply:
Yes, v2 is optional and v1 will always be there unless customer configures for v2 only. It's not up to the plugin to determine the configuration, therefore I feel that both v2 and v1 failure can be debug statement and not getting any metadata can be a warning statement.

Reply:
Yes. v2 is optional and v1 will always be present unless customer has configured v2 only. I think it would be better to have debugs for v2 or v1 failure and a warning for not being able to fetch the metadata at all.

Reply:
Added debugs and a warning